### PR TITLE
Added --type flag #1749

### LINKED ISSUE #1749: Additional flag to select the types of IaC used for the repository, so only queries for those types will run - performance improvement
**Is your feature request related to a problem? Please describe.**
In order to run only the required queries, user should be able to provide a flag specifying the types of IaC included in the source code, so only queries for those types will run.

**Describe the solution you'd like**
Additional flag for kics, for example: --type "k8s,docker" -> only queries of kubernetes and dockerfiles will run this way, discarding the other ones - this will increase the performance of kics and reduce scan time

### PR DESCRIPTION
**Pull request by joaoReigota1**
Closes #1749

**Proposed Changes**

- Added --type flag to specify which type of queries to use in the scan
- --type will restrict parser to only use parsers of the given types

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -947,6 +947,7 @@ github.com/stretchr/objx v0.0.0-20180129172003-8a3f7159479f/go.mod h1:HFkY916IF+
 github.com/stretchr/objx v0.1.0/go.mod h1:HFkY916IF+rwdDfMAkV7OtwuqBVzrE8GR6GFx+wExME=
 github.com/stretchr/objx v0.1.1 h1:2vfRuCMp5sSVIDSqO8oNnWJq7mPa6KVP3iPIwFBuy8A=
 github.com/stretchr/objx v0.1.1/go.mod h1:HFkY916IF+rwdDfMAkV7OtwuqBVzrE8GR6GFx+wExME=
+github.com/stretchr/objx v0.2.0 h1:Hbg2NidpLE8veEBkEZTL3CvlkUIVzuU9jDplZO54c48=
```
Comment (rogeriopeixotocx) [inline go.sum:950]: do we need to change the go.sum?

File: internal/console/scan.go [hunk 2]
```diff
@@ -97,11 +104,10 @@ func scan() error {
 	}
 
 	combinedParser := parser.NewBuilder().
-		// Add(&jsonParser.Parser{}).
```
Comment (rogeriopeixotocx) [inline internal/console/scan.go:100]: thank you for removing this :D

File: pkg/engine/query/source.go [hunk 3]
```diff
@@ -71,21 +71,24 @@ func (s *FilesystemSource) GetGenericQuery(platform string) (string, error) {
...
-	err := filepath.Walk(s.Source,
-		func(p string, f os.FileInfo, err error) error {
-			if err != nil {
-				return err
-			}
-
-			if f.IsDir() || f.Name() != QueryFileName {
+	var err error
+	for _, path := range s.Source {
+		err = filepath.Walk(path,
```
Comment (rogeriopeixotocx) [inline pkg/engine/query/source.go:76]: any performance concerns in using Walk inside a loop?

File: test/fixtures/type-test01/template01/metadata.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+  "id": "<ID>",
+  "queryName": "<QUERY_NAME>",
+  "severity": "HIGH",
+  "category": null,
+  "descriptionText": "<TEXT>",
+  "descriptionUrl": "#"
+}
```
Comment (rogeriopeixotocx) [inline test/fixtures/type-test01/template01/metadata.json:8]: EOF newline missing

File: test/fixtures/type-test01/template01/query.rego [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+  resource == "<VALUE>"
+
+	result := {
+                "documentId": 		input.document[i].id,
+                "searchKey": 	    sprintf("%s", [resource]),
+                "issueType":		"IncorrectValue",  #"MissingAttribute" / "RedundantAttribute"
+                "keyExpectedValue": "<RESOURCE>",
+                "keyActualValue": 	resource
+              }
+}
```
Comment (rogeriopeixotocx) [inline test/fixtures/type-test01/template01/query.rego:14]: EOF newline missing

File: test/fixtures/type-test01/template01/test/positive.tf [hunk 6]
```diff
@@ -0,0 +1 @@
+#this is a problematic code where the query should report a result(s)
```
Comment (rogeriopeixotocx) [inline test/fixtures/type-test01/template01/test/positive.tf:1]: EOF newline missing

File: test/fixtures/type-test01/template01/test/negative.tf [hunk 7]
```diff
@@ -0,0 +1 @@
+#this code is a correct code for which the query should not find any result
```
Comment (rogeriopeixotocx) [inline test/fixtures/type-test01/template01/test/negative.tf:1]: EOF newline missing

File: test/fixtures/type-test01/template02/metadata.json [hunk 8]
```diff
@@ -0,0 +1,8 @@
+{
+  "id": "<ID>",
+  "queryName": "<QUERY_NAME>",
+  "severity": "HIGH",
+  "category": null,
+  "descriptionText": "<TEXT>",
+  "descriptionUrl": "#"
+}
```
Comment (rogeriopeixotocx) [inline test/fixtures/type-test01/template02/metadata.json:8]: EOF newline missing

File: test/fixtures/type-test01/template02/query.rego [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+  resource == "<VALUE>"
+
+	result := {
+                "documentId": 		input.document[i].id,
+                "searchKey": 	    sprintf("%s", [resource]),
+                "issueType":		"IncorrectValue",  #"MissingAttribute" / "RedundantAttribute"
+                "keyExpectedValue": "<RESOURCE>",
+                "keyActualValue": 	resource
+              }
+}
```
Comment (rogeriopeixotocx) [inline test/fixtures/type-test01/template02/query.rego:14]: EOF newline missing

File: test/fixtures/type-test01/template02/test/negative.tf [hunk 10]
```diff
@@ -0,0 +1 @@
+#this code is a correct code for which the query should not find any result
```
Comment (rogeriopeixotocx) [inline test/fixtures/type-test01/template02/test/negative.tf:1]: EOF newline missing

File: test/fixtures/type-test01/template02/test/positive.tf [hunk 11]
```diff
@@ -0,0 +1 @@
+#this is a problematic code where the query should report a result(s)
```
Comment (rogeriopeixotocx) [inline test/fixtures/type-test01/template02/test/positive.tf:1]: EOF newline missing

File: test/fixtures/type-test01/template03/metadata.json [hunk 12]
```diff
@@ -0,0 +1,8 @@
+{
+  "id": "<ID>",
+  "queryName": "<QUERY_NAME>",
+  "severity": "HIGH",
+  "category": null,
+  "descriptionText": "<TEXT>",
+  "descriptionUrl": "#"
+}
```
Comment (rogeriopeixotocx) [inline test/fixtures/type-test01/template03/metadata.json:8]: EOF newline missing

File: test/fixtures/type-test01/template03/query.rego [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+  resource == "<VALUE>"
+
+	result := {
+                "documentId": 		input.document[i].id,
+                "searchKey": 	    sprintf("%s", [resource]),
+                "issueType":		"IncorrectValue",  #"MissingAttribute" / "RedundantAttribute"
+                "keyExpectedValue": "<RESOURCE>",
+                "keyActualValue": 	resource
+              }
+}
```
Comment (rogeriopeixotocx) [inline test/fixtures/type-test01/template03/query.rego:14]: EOF newline missing

File: test/fixtures/type-test01/template03/test/negative.tf [hunk 14]
```diff
@@ -0,0 +1 @@
+#this code is a correct code for which the query should not find any result
```
Comment (rogeriopeixotocx) [inline test/fixtures/type-test01/template03/test/negative.tf:1]: EOF newline missing

File: test/fixtures/type-test01/template03/test/positive.tf [hunk 15]
```diff
@@ -0,0 +1 @@
+#this is a problematic code where the query should report a result(s)
```
Comment (rogeriopeixotocx) [inline test/fixtures/type-test01/template03/test/positive.tf:1]: EOF newline missing

File: test/fixtures/type-test02/template01/metadata.json [hunk 16]
```diff
@@ -0,0 +1,8 @@
+{
+  "id": "<ID>",
+  "queryName": "<QUERY_NAME>",
+  "severity": "HIGH",
+  "category": null,
+  "descriptionText": "<TEXT>",
+  "descriptionUrl": "#"
+}
```
Comment (rogeriopeixotocx) [inline test/fixtures/type-test02/template01/metadata.json:8]: EOF newline missing

File: test/fixtures/type-test02/template01/query.rego [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+  resource == "<VALUE>"
+
+	result := {
+                "documentId": 		input.document[i].id,
+                "searchKey": 	    sprintf("%s", [resource]),
+                "issueType":		"IncorrectValue",  #"MissingAttribute" / "RedundantAttribute"
+                "keyExpectedValue": "<RESOURCE>",
+                "keyActualValue": 	resource
+              }
+}
```
Comment (rogeriopeixotocx) [inline test/fixtures/type-test02/template01/query.rego:14]: EOF newline missing

File: test/fixtures/type-test02/template01/test/negative.tf [hunk 18]
```diff
@@ -0,0 +1 @@
+#this code is a correct code for which the query should not find any result
```
Comment (rogeriopeixotocx) [inline test/fixtures/type-test02/template01/test/negative.tf:1]: EOF newline missing

File: test/fixtures/type-test02/template01/test/positive.tf [hunk 19]
```diff
@@ -0,0 +1 @@
+#this is a problematic code where the query should report a result(s)
```
Comment (rogeriopeixotocx) [inline test/fixtures/type-test02/template01/test/positive.tf:1]: EOF newline missing

File: test/fixtures/type-test02/template02/metadata.json [hunk 20]
```diff
@@ -0,0 +1,8 @@
+{
+  "id": "<ID>",
+  "queryName": "<QUERY_NAME>",
+  "severity": "HIGH",
+  "category": null,
+  "descriptionText": "<TEXT>",
+  "descriptionUrl": "#"
+}
```
Comment (rogeriopeixotocx) [inline test/fixtures/type-test02/template02/metadata.json:8]: EOF newline missing

File: test/fixtures/type-test02/template02/query.rego [hunk 21]
```diff
@@ -0,0 +1,14 @@
...
+  resource == "<VALUE>"
+
+	result := {
+                "documentId": 		input.document[i].id,
+                "searchKey": 	    sprintf("%s", [resource]),
+                "issueType":		"IncorrectValue",  #"MissingAttribute" / "RedundantAttribute"
+                "keyExpectedValue": "<RESOURCE>",
+                "keyActualValue": 	resource
+              }
+}
```
Comment (rogeriopeixotocx) [inline test/fixtures/type-test02/template02/query.rego:14]: EOF newline missing

File: test/fixtures/type-test02/template02/test/negative.tf [hunk 22]
```diff
@@ -0,0 +1 @@
+#this code is a correct code for which the query should not find any result
```
Comment (rogeriopeixotocx) [inline test/fixtures/type-test02/template02/test/negative.tf:1]: EOF newline missing

File: test/fixtures/type-test02/template02/test/positive.tf [hunk 23]
```diff
@@ -0,0 +1 @@
+#this is a problematic code where the query should report a result(s)
```
Comment (rogeriopeixotocx) [inline test/fixtures/type-test02/template02/test/positive.tf:1]: EOF newline missing

**Review by rogeriopeixotocx — CHANGES_REQUESTED** [review]
Other than minor fixes and questions I believe it looks good.

**Comment by rogeriopeixotocx** [comment]
@ruigomescx @joaoReigota1  should we create a benchmark test to the `source.GetQueries()` function to check performance for `filepath.Walk()` inside a loop?

File: internal/console/scan.go (hunk 2 above)
Comment (joaoReigota1) [reply internal/console/scan.go:100]: :D

File: go.sum (hunk 1 above)
Comment (joaoReigota1) [reply go.sum:950]: It doesn't I will revert

File: test/fixtures/type-test01/template01/metadata.json (hunk 4 above)
Comment (joaoReigota1) [reply test/fixtures/type-test01/template01/metadata.json]: resolved

File: test/fixtures/type-test02/template01/test/positive.tf (hunk 19 above)
Comment (joaoReigota1) [reply test/fixtures/type-test02/template01/test/positive.tf]: resolved

File: test/fixtures/type-test02/template02/query.rego (hunk 21 above)
Comment (joaoReigota1) [reply test/fixtures/type-test02/template02/query.rego]: resolved

File: test/fixtures/type-test02/template02/test/positive.tf (hunk 23 above)
Comment (joaoReigota1) [reply test/fixtures/type-test02/template02/test/positive.tf]: resolved

File: test/fixtures/type-test01/template01/test/positive.tf (hunk 6 above)
Comment (joaoReigota1) [reply test/fixtures/type-test01/template01/test/positive.tf]: resolved

File: test/fixtures/type-test01/template01/query.rego (hunk 5 above)
Comment (joaoReigota1) [reply test/fixtures/type-test01/template01/query.rego]: resolved

File: test/fixtures/type-test01/template02/metadata.json (hunk 8 above)
Comment (joaoReigota1) [reply test/fixtures/type-test01/template02/metadata.json]: resolved

File: test/fixtures/type-test01/template01/test/negative.tf (hunk 7 above)
Comment (joaoReigota1) [reply test/fixtures/type-test01/template01/test/negative.tf]: resolved

File: test/fixtures/type-test01/template02/query.rego (hunk 9 above)
Comment (joaoReigota1) [reply test/fixtures/type-test01/template02/query.rego]: resolved

File: test/fixtures/type-test01/template02/test/positive.tf (hunk 11 above)
Comment (joaoReigota1) [reply test/fixtures/type-test01/template02/test/positive.tf]: resolved

File: test/fixtures/type-test01/template02/test/negative.tf (hunk 10 above)
Comment (joaoReigota1) [reply test/fixtures/type-test01/template02/test/negative.tf]: resolved

File: test/fixtures/type-test01/template03/metadata.json (hunk 12 above)
Comment (joaoReigota1) [reply test/fixtures/type-test01/template03/metadata.json]: resolved

File: test/fixtures/type-test01/template03/query.rego (hunk 13 above)
Comment (joaoReigota1) [reply test/fixtures/type-test01/template03/query.rego]: resolved

File: test/fixtures/type-test01/template03/test/negative.tf (hunk 14 above)
Comment (joaoReigota1) [reply test/fixtures/type-test01/template03/test/negative.tf]: resolved

File: test/fixtures/type-test01/template03/test/positive.tf (hunk 15 above)
Comment (joaoReigota1) [reply test/fixtures/type-test01/template03/test/positive.tf]: resolved

File: test/fixtures/type-test02/template01/metadata.json (hunk 16 above)
Comment (joaoReigota1) [reply test/fixtures/type-test02/template01/metadata.json]: resolved

File: test/fixtures/type-test02/template01/query.rego (hunk 17 above)
Comment (joaoReigota1) [reply test/fixtures/type-test02/template01/query.rego]: resolved

File: test/fixtures/type-test02/template01/test/negative.tf (hunk 18 above)
Comment (joaoReigota1) [reply test/fixtures/type-test02/template01/test/negative.tf]: resolved

File: test/fixtures/type-test02/template02/metadata.json (hunk 20 above)
Comment (joaoReigota1) [reply test/fixtures/type-test02/template02/metadata.json]: resolved

File: test/fixtures/type-test02/template02/test/negative.tf (hunk 22 above)
Comment (joaoReigota1) [reply test/fixtures/type-test02/template02/test/negative.tf]: resolved

**Comment by rogeriopeixotocx** [comment]
@joaoReigota1 resolve conflicts, please

**Comment by rogeriopeixotocx** [comment]
@joaoReigota1 fix conflicts again, please :)

**Review by rogeriopeixotocx — APPROVED** [review]
LGTM

File: assets/queries/template/metadata.json [hunk 24]
```diff
@@ -4,5 +4,6 @@
   "severity": "HIGH",
   "category": null,
   "descriptionText": "<TEXT>",
-  "descriptionUrl": "#"
-}
\ No newline at end of file
+  "descriptionUrl": "#",
+  "platform": "template"
```
Comment (felipe-avelar) [inline assets/queries/template/metadata.json:8]: Wouldn't be better if it was &lt;PLATFORM&gt; instead of template?

**Review by ruigomescx — DISMISSED** [review]

**Review by felipe-avelar — COMMENTED** [review]
Check suggested changes, please

File: assets/queries/template/metadata.json (hunk 24 above)
Comment (joaoReigota1) [reply assets/queries/template/metadata.json]: resolved

**Review by felipe-avelar — APPROVED** [review]
LGTM

**Review by felipe-avelar — CHANGES_REQUESTED** [review]
Check in windows environment

**Review by felipe-avelar — APPROVED** [review]
LGTM

**Review by felipe-avelar — APPROVED** [review]
LGTM

**Review by ruigomescx — APPROVED** [review]

**Pull request merged**